airflow/datasets/__init__.py

```diff
@@ -106,8 +112,22 @@ class DatasetAny(_DatasetBooleanCondition):
 
     agg_func = any
 
+    def __or__(self, other: BaseDatasetEventInput) -> DatasetAny:
+        # Optimization: X | (Y | Z) is equivalent to X | Y | Z.
```

Comment:
@sunank200 i think you meant to mark this as a todo?

i.e.
```
# todo: Optimization: X | (Y | Z) is equivalent to X | Y | Z.
```

i.e. should just tweak the language to make clear this is a _potential_ optimization, not one that is already committed that you are simply telling us about

Reply:
alternatively if you don't want to think of it as a "to do" you could write
```
# potential optimization: X | (Y | Z) is equivalent to X | Y | Z.
```

Reply:
I think this was last minute commit from TP: https://github.com/apache/airflow/pull/37101/commits/2d670847d74790eee482c2f684a8ab0e0ddb0bec I can do the change as required.

Reply:
TP had commented more context to this change here: https://github.com/apache/airflow/pull/37101#pullrequestreview-1900009726

Reply:
I messed up the example though, the thing we’re optimising here is `(X | Y) | Z` instead. Let’s fix this.